java/vector/src/main/java/org/apache/arrow/vector/stream/ArrowStreamWriter.java

```diff
@@ -0,0 +1,135 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.arrow.vector.stream;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.channels.WritableByteChannel;
+import java.util.List;
+
+import org.apache.arrow.vector.file.WriteChannel;
+import org.apache.arrow.vector.schema.ArrowBuffer;
+import org.apache.arrow.vector.schema.ArrowRecordBatch;
+import org.apache.arrow.vector.types.pojo.Schema;
+
+import io.netty.buffer.ArrowBuf;
+
+public class ArrowStreamWriter implements AutoCloseable {
+  private final WriteChannel out;
+  private final ArrowStreamHeader header;
+  private boolean headerSent = false;
+
+  /**
+   * Creates the stream writer. non-blocking.
+   * totalBatches can be set if the writer knows beforehand. Can be -1 if unknown.
+   */
+  public ArrowStreamWriter(WritableByteChannel out, Schema schema, int totalBatches) {
+    this.out = new WriteChannel(out);
+    this.header = new ArrowStreamHeader(schema, totalBatches);
+  }
+
+  public ArrowStreamWriter(OutputStream out, Schema schema, int totalBatches)
+      throws IOException {
+    this(Channels.newChannel(out), schema, totalBatches);
+  }
+
+  public long bytesWritten() { return out.getCurrentPosition(); }
+
+  /**
+   * Computes the size of the serialized body for this recordBatch.
+   */
+  public int computeBodyLength(ArrowRecordBatch recordBatch) {
+    int size = 0;
+
+    List<ArrowBuf> buffers = recordBatch.getBuffers();
+    List<ArrowBuffer> buffersLayout = recordBatch.getBuffersLayout();
+    if (buffers.size() != buffersLayout.size()) {
+      throw new IllegalStateException("the layout does not match: " +
+          buffers.size() + " != " + buffersLayout.size());
+    }
+
+    for (int i = 0; i < buffers.size(); i++) {
+      ArrowBuf buffer = buffers.get(i);
+      ArrowBuffer layout = buffersLayout.get(i);
+      size += (layout.getOffset() - size);
+      ByteBuffer nioBuffer =
+          buffer.nioBuffer(buffer.readerIndex(), buffer.readableBytes());
+      size += nioBuffer.remaining();
+    }
+    return size;
+  }
```

Comment:
Yea, it could also make sense to consolidate all of this into org.apache.arrow.vector.ipc instead of two packages as well. Similar comment as above about a follow up patch.